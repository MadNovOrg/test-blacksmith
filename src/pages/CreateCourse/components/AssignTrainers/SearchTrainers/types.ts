import type {
  Profile,
  InviteStatus,
  CourseSchedule,
  SearchTrainerAvailability,
} from '@app/types'

export type SearchTrainer = {
  availability?: SearchTrainerAvailability
} & Pick<Profile, 'id' | 'fullName' | 'avatar'>

export type SearchTrainersSchedule = Partial<
  Pick<CourseSchedule, 'start' | 'end'>
>

export type SearchTrainerBookings = {
  profile_id: string
  status: InviteStatus
  course: {
    schedule: Array<{
      start: string
      end: string
    }>
  }
}
