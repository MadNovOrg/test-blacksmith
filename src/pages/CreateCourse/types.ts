import {
  CourseType,
  Draft,
  ExpensesInput,
  InvoiceDetails,
  TrainerInput,
  ValidCourseInput,
} from '@app/types'

import { CourseException } from './components/CourseExceptionsConfirmation/utils'
import { SaveDraftResult } from './components/CreateCourseProvider'

export enum StepsEnum {
  COURSE_DETAILS = 'course-details',
  ASSIGN_TRAINER = 'assign-trainer',
  TRAINER_EXPENSES = 'trainer-expenses',
  LICENSE_ORDER_DETAILS = 'license-order-details',
  ORDER_DETAILS = 'order-details',
  REVIEW_AND_CONFIRM = 'review-and-confirm',
  COURSE_BUILDER = 'course-builder',
}

export type ContextValue = {
  completeStep: (step: StepsEnum) => void
  completedSteps: StepsEnum[]
  courseData?: ValidCourseInput
  courseName: string
  pricing: {
    amount?: number
    error?: boolean
  }
  courseType: CourseType
  currentStepKey: StepsEnum | null
  draftName?: string | null
  expenses: Record<string, ExpensesInput>
  saveDraft: (name?: string) => Promise<SaveDraftResult>
  setCourseData: (courseData: ValidCourseInput) => void
  setCurrentStepKey: (step: StepsEnum) => void
  setExpenses: (expenses: Record<string, ExpensesInput>) => void
  setTrainers: (trainers: TrainerInput[]) => void
  trainers: TrainerInput[]
  setGo1Licensing: (go1Licensing: Draft['go1Licensing']) => void
  setInvoiceDetails: (invoiceDetails: InvoiceDetails) => void
  go1Licensing: Draft['go1Licensing']
  exceptions: CourseException[]
  invoiceDetails?: InvoiceDetails
  showDraftConfirmationDialog: boolean
  setShowDraftConfirmationDialog: (value: boolean) => void
}
