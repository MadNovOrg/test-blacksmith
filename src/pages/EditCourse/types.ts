export enum RescheduleFeeType {
  APPLY_TERMS = 'APPLY_TERMS',
  CUSTOM_FEE = 'CUSTOM_FEE',
  NO_FEE = 'NO_FEE',
}

export type CourseDiff = {
  type: 'date' | 'venue'
  oldValue: string | Date[]
  newValue: string | Date[]
}
