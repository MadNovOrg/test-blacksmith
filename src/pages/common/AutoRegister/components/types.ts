import { schemas, yup, TFunction } from '@app/schemas'
import { requiredMsg } from '@app/util'

export type FormInputs = {
  firstName: string
  surname: string
  marketing: boolean
  phone: string
  password: string
  dob: Date | null
  tcs: boolean
}

export const getFormSchema = (t: TFunction) => {
  return yup.object({
    firstName: yup.string().required(requiredMsg(t, 'first-name')),
    surname: yup.string().required(requiredMsg(t, 'surname')),
    password: schemas.password(t),

    phone: yup.string().required(requiredMsg(t, 'phone')),
    dob: yup
      .date()
      .typeError(t('validation-errors.invalid-date-optional'))
      .nullable(),

    marketing: yup.boolean(),
    tcs: yup.boolean().oneOf([true], t('pages.signup.tcs-required')),
  })
}
