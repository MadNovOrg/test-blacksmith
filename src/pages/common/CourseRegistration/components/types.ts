import { schemas, yup, TFunction } from '@app/schemas'
import { requiredMsg } from '@app/util'

import { sectors } from './org-data'

type Sector = keyof typeof sectors

export type FormInputs = {
  firstName: string
  surname: string
  email: string
  marketing: boolean
  countryCode: string
  phone: string
  password: string
  confirmPassword: string
  dob: Date | null
  orgId: string | null
  sector: Sector | ''
  position: string
  otherPosition: string
  tcs: boolean
}

export const getFormSchema = (t: TFunction) => {
  return yup.object({
    firstName: yup.string().required(requiredMsg(t, 'first-name')),

    surname: yup.string().required(requiredMsg(t, 'surname')),

    email: schemas.email(t),
    marketing: yup.boolean(),

    countryCode: yup.string(),
    phone: yup.string(),
    password: schemas.password(t),

    dob: yup.date(),
    orgId: yup
      .string()
      .required(requiredMsg(t, 'org-name'))
      .typeError(requiredMsg(t, 'org-name')),
    sector: yup.string().required(),
    position: yup.string().required(),
    otherPosition: yup.string().when('position', {
      is: 'other',
      then: yup
        .string()
        .required(t('validation-errors.other-position-required')),
    }),

    tcs: yup.boolean().oneOf([true], t('pages.signup.tcs-required')),
  })
}
