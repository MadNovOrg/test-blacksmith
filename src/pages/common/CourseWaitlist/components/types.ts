import { schemas, yup, TFunction } from '@app/schemas'
import { requiredMsg } from '@app/util'

export const getFormSchema = (t: TFunction) => {
  return yup.object({
    firstName: yup.string().required(requiredMsg(t, 'first-name')),

    surname: yup.string().required(requiredMsg(t, 'surname')),

    email: schemas.email(t, true),

    phone: yup
      .string()
      .required(requiredMsg(t, 'phone'))
      .min(10, t('validation-errors.phone-invalid')),

    orgName: yup.string().required(requiredMsg(t, 'org-name')),
  })
}
