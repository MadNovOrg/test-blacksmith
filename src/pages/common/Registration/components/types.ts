import { schemas, yup, TFunction } from '@app/schemas'
import { requiredMsg } from '@app/util'

export type FormInputs = {
  firstName: string
  surname: string
  email: string
  marketing: boolean
  phone: string
  password: string
  dob: Date | null
  tcs: boolean
}

export const getFormSchema = (t: TFunction) => {
  return yup.object({
    firstName: yup.string().required(requiredMsg(t, 'first-name')),

    surname: yup.string().required(requiredMsg(t, 'surname')),

    email: schemas.email(t),
    marketing: yup.boolean(),

    phone: yup.string().required(requiredMsg(t, 'phone')),
    password: schemas.password(t),

    dob: yup.date().typeError(t('validation-errors.invalid-date')),

    tcs: yup.boolean().oneOf([true], t('pages.signup.tcs-required')),
  })
}
