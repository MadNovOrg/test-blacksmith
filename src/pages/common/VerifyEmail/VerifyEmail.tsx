import { Box, Typography, Button } from '@mui/material'
import { Auth } from 'aws-amplify'
import { TFunction } from 'i18next'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import { useLocation, useNavigate } from 'react-router-dom'

import { UnverifiedLayout } from '@app/components/UnverifiedLayout'
import { useAuth } from '@app/context/auth'
import { schemas, yup } from '@app/schemas'
import { requiredMsg } from '@app/util'

import { Form } from './components/Form'

type LocationState = {
  from: { pathname: string; search: string }
  send: boolean
}

export type Props = unknown

export type VerifyInputs = { code: string }

export const getVerifySchema = (t: TFunction) => {
  return yup.object({
    code: schemas
      .emailCode(t)
      .required(requiredMsg(t, 'pages.signup.verify-code-label')),
  })
}

export const VerifyEmailPage: React.FC<Props> = () => {
  const { t } = useTranslation()
  const location = useLocation()
  const { loadProfile } = useAuth()
  const navigate = useNavigate()
  const [success, setSuccess] = useState(false)
  const state = (location.state || {}) as LocationState
  const from = state.from || {}

  const handleContinue = async () => {
    const currentUser = await Auth.currentUserPoolUser()
    await currentUser.refreshSessionIfPossible()
    await loadProfile(currentUser)

    const to = `${from.pathname || '/'}${from.search || ''}`
    return navigate(to, { replace: true })
  }

  return (
    <UnverifiedLayout width={628}>
      <Typography
        variant="h3"
        sx={{ textAlign: 'center', fontWeight: 600, color: 'grey.800' }}
      >
        {t('pages.signup.verify-heading')}
      </Typography>

      {success ? (
        <Box>
          <Typography sx={{ my: 6 }}>
            {t('pages.signup.success-hint')}
          </Typography>

          <Button
            variant="contained"
            color="primary"
            size="large"
            onClick={handleContinue}
            data-testid="btn-goto-login"
          >
            {t('pages.signup.success-btn')}
          </Button>
        </Box>
      ) : (
        <Form
          onVerifyLater={() => navigate('/')}
          onSuccess={() => setSuccess(true)}
        />
      )}
    </UnverifiedLayout>
  )
}
