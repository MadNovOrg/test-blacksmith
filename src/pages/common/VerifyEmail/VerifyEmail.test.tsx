import { Auth } from 'aws-amplify'
import React from 'react'
import { Route, Routes } from 'react-router-dom'

import { useAuth } from '@app/context/auth'
import { AuthContextType } from '@app/context/auth/types'

import {
  render,
  screen,
  userEvent,
  waitForCalls,
  waitFor,
  waitForText,
} from '@test/index'

import { VerifyEmailPage } from './VerifyEmail'

jest.mock('@app/context/auth', () => ({
  ...jest.requireActual('@app/context/auth'),
  useAuth: jest.fn().mockReturnValue({ loadProfile: jest.fn() }),
}))

jest.mock('aws-amplify', () => ({
  Auth: {
    verifyCurrentUserAttribute: jest.fn().mockResolvedValue({}),
    verifyCurrentUserAttributeSubmit: jest.fn().mockResolvedValue({}),
    currentUserPoolUser: jest.fn(),
  },
}))
const AuthMock = jest.mocked(Auth)
const useAuthMock = jest.mocked(useAuth)

describe('page: VerifyEmailPage', () => {
  const setup = () => {
    return render(
      <Routes>
        <Route path="/" element={<VerifyEmailPage />} />
        <Route path="/profile" element={<div />} />
      </Routes>,
      {},
      { initialEntries: ['/'] }
    )
  }

  it('sends verification email when verify now is clicked', async () => {
    render(<VerifyEmailPage />)

    expect(screen.queryByTestId('signup-verify-btn')).not.toBeInTheDocument()
    await userEvent.click(screen.getByTestId('signup-verify-now-btn'))
    await waitForCalls(AuthMock.verifyCurrentUserAttribute)
    expect(AuthMock.verifyCurrentUserAttribute).toHaveBeenCalledWith('email')
    await waitFor(() =>
      expect(screen.queryByTestId('signup-verify-btn')).toBeInTheDocument()
    )
  })

  it('does not submit when code is empty', async () => {
    render(<VerifyEmailPage />)

    await userEvent.click(screen.getByTestId('signup-verify-now-btn'))
    await waitFor(() =>
      expect(screen.queryByTestId('signup-verify-btn')).toBeInTheDocument()
    )

    expect(screen.queryByTestId('signup-verify-error')).toBeNull()

    await userEvent.click(screen.getByTestId('signup-verify-btn'))

    await waitForText('Please enter 6 digit passcode received in email')
  })

  it('shows error when code is invalid', async () => {
    setup()

    await userEvent.click(screen.getByTestId('signup-verify-now-btn'))
    await waitFor(() =>
      expect(screen.queryByTestId('signup-verify-btn')).toBeInTheDocument()
    )

    const code = '1234'
    await fillCode(code)

    const submitBtn = screen.getByTestId('signup-verify-btn')
    await userEvent.click(submitBtn)

    const error = await screen.findByTestId('signup-verify-error')
    expect(error).toHaveTextContent(
      'Please enter 6 digit passcode received in email'
    )

    expect(AuthMock.verifyCurrentUserAttributeSubmit).not.toHaveBeenCalled()
  })

  it('calls cognito submit when code is valid', async () => {
    AuthMock.verifyCurrentUserAttributeSubmit.mockResolvedValue('')

    setup()

    await userEvent.click(screen.getByTestId('signup-verify-now-btn'))
    await waitFor(() =>
      expect(screen.queryByTestId('signup-verify-btn')).toBeInTheDocument()
    )

    const code = '123456'
    await fillCode(code)

    const submitBtn = screen.getByTestId('signup-verify-btn')
    await userEvent.click(submitBtn)

    await waitForCalls(AuthMock.verifyCurrentUserAttributeSubmit)
    expect(AuthMock.verifyCurrentUserAttributeSubmit).toHaveBeenCalledWith(
      'email',
      code
    )
    await waitFor(() =>
      expect(screen.queryByTestId('btn-goto-login')).toBeInTheDocument()
    )
  })

  it('shows error when code is incorrect', async () => {
    AuthMock.verifyCurrentUserAttributeSubmit.mockRejectedValue({
      code: 'CodeMismatchException',
    })

    setup()

    await userEvent.click(screen.getByTestId('signup-verify-now-btn'))
    await waitFor(() =>
      expect(screen.queryByTestId('signup-verify-btn')).toBeInTheDocument()
    )

    const code = '123456'
    await fillCode(code)

    const submitBtn = screen.getByTestId('signup-verify-btn')
    await userEvent.click(submitBtn)

    await waitForCalls(AuthMock.verifyCurrentUserAttributeSubmit)

    const error = await screen.findByTestId('signup-verify-error')
    expect(error).toHaveTextContent('Verification code is incorrect.')
  })

  it('shows error when code is expired', async () => {
    AuthMock.verifyCurrentUserAttributeSubmit.mockRejectedValue({
      code: 'ExpiredCodeException',
    })

    setup()

    await userEvent.click(screen.getByTestId('signup-verify-now-btn'))
    await waitFor(() =>
      expect(screen.queryByTestId('signup-verify-btn')).toBeInTheDocument()
    )

    const code = '123456'

    await fillCode(code)

    const submitBtn = screen.getByTestId('signup-verify-btn')
    userEvent.click(submitBtn)

    await waitForCalls(AuthMock.verifyCurrentUserAttributeSubmit)

    const error = await screen.findByTestId('signup-verify-error')
    expect(error).toHaveTextContent(
      'Verification code expired. Please request new code.'
    )
  })

  it('sends new code when resend is pressed', async () => {
    setup()

    await userEvent.click(screen.getByTestId('signup-verify-now-btn'))
    await waitFor(() =>
      expect(screen.queryByTestId('signup-verify-btn')).toBeInTheDocument()
    )

    const resendBtn = screen.getByTestId('signup-verify-resend')
    await userEvent.click(resendBtn)

    await waitForCalls(AuthMock.verifyCurrentUserAttribute, 2)
    expect(AuthMock.verifyCurrentUserAttribute).toHaveBeenCalledWith('email')
  })

  it('reloads profile before navigating away', async () => {
    AuthMock.verifyCurrentUserAttributeSubmit.mockResolvedValue('')

    const refreshSessionIfPossibleMock = jest.fn().mockResolvedValue('')
    AuthMock.currentUserPoolUser.mockResolvedValue({
      refreshSessionIfPossible: refreshSessionIfPossibleMock,
    })

    const loadProfileMock = jest.fn().mockResolvedValue('')
    useAuthMock.mockReturnValue({
      loadProfile: loadProfileMock,
    } as unknown as AuthContextType)

    setup()

    await userEvent.click(screen.getByTestId('signup-verify-now-btn'))
    await waitFor(() =>
      expect(screen.queryByTestId('signup-verify-btn')).toBeInTheDocument()
    )

    const code = '123456'

    await fillCode(code)

    const submitBtn = screen.getByTestId('signup-verify-btn')
    await userEvent.click(submitBtn)

    await waitFor(() =>
      expect(screen.queryByTestId('btn-goto-login')).toBeInTheDocument()
    )

    await userEvent.click(screen.getByTestId('btn-goto-login'))

    await waitForCalls(AuthMock.currentUserPoolUser)
    await waitForCalls(refreshSessionIfPossibleMock)
    await waitForCalls(loadProfileMock)
  })
})

async function fillCode(code: string) {
  await Promise.all(
    code.split('').map((n, i) => {
      const iN = screen.getByTestId(`signup-verify-code-${i}`)
      return userEvent.type(iN, n)
    })
  )
}
