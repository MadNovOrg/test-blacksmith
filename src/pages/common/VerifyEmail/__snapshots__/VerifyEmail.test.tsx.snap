// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page: VerifyEmailPage matchs snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-16d3e73"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionFixed mui-fixed css-1jvst5f-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1k7giva-MuiToolbar-root"
      >
        <div
          class="MuiBox-root css-0"
        />
        <div
          class="MuiBox-root css-70qvj9"
        >
          <div
            class="MuiBox-root css-70qvj9"
          >
            <div
              data-testid="app-logo"
              height="48"
              width="230"
            />
          </div>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <a
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-kboeec-MuiButtonBase-root-MuiButton-root"
            href="/verify"
            tabindex="0"
          >
            Verify
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </a>
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root  css-kboeec-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Logout
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
    <div
      class="MuiBox-root css-1wlfb0w"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-2w99rx-MuiTypography-root"
      >
        Email Verification
      </h3>
      <form
        class="MuiBox-root css-dvxtzn"
      >
        <button
          class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root css-o5kj1d-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
          data-testid="signup-verify-now-btn"
          id="mui-1"
          tabindex="0"
          type="button"
        >
          Verify Now
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeLarge MuiButton-textSizeLarge MuiButtonBase-root  css-e432cs-MuiButtonBase-root-MuiButton-root"
          data-testid="signup-verify-later-btn"
          tabindex="0"
          type="button"
        >
          Verify Later
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </div>
</div>
`;
